### ISSUE TITLE
Monitoring etcd created by CAPI using prometheus-operator

### ISSUE THREAD

**Issue by MPV** [kubernetes-sigs/cluster-api#3939]
Anyone who has set up monitoring of the `etcd` created by CAPI using `prometheus-operator`?

How have you created the necessary certs (as k8s secrets into the target cluster) for the Prometheus instance/ServiceMonitor to use to auth with etcd when scraping its metrics port?

I'd like to automate something like this when creating my new clusters with CAPI:

https://jpweber.io/blog/monitoring-external-etcd-cluster-with-prometheus-operator/

Is creating a `ResourceSet` the way to go here?

**Comment by MPV** [comment]
Or would using `postKubeadmCommands` make more sense?

**Comment by vincepri** [comment]
/triage support
/milestone Next

cc @randomvariable @fabriziopandini

**Comment by fabriziopandini** [comment]
I don't know in detail how the Prometheus operator works, but from a quick look at the article there are several points to be considered:
- Solution will be different if using a stacked etcd or an external etcd. For now, I focus on the first scenario only.
- Defining the list of Endpoints requires knowing in advance machine IPs. Most probably this should be changed in order to use a selector that targets etcd pods.
- ServiceMonitor monitor requires some path, but it is not clear to me where those files should come from (the machine? the target etcd pod). In CAPI, we are creating secrets inside the etcd pod, and an option is to re-use some of them for scraping (e.g. apiserver-etcd-client)
- If instead it is required to generate dedicate certs, the way forward is to provide a custom etcd-ca before the cluster gets create, provide the additional certs to the machines, and probably customize the cluster configuration by adding etcd extra volumes

TL;DR with the caveat of a general lack of background info about the Prometheus operator, IMO there are two options to explore for trying to use the Prometheus operator with the stacked etcd cluster created by CAPI:
A) reusing existing certs already available inside the etcd pod
B) providing a custom etcd CA before creating the cluster and inject additional etcd certs into machines/etcd pods

**Comment by randomvariable** [comment]
My only issue with reusing the client certs as per the blog post is that it leaves Prometheus a bit overpowered in terms of permissions and could be a security risk.

There's a number of overlapping problems here:
kubeadm configures etcd to use the same CA for both standard client certificates and etcd peer certificates. Unfortunately, this means any valid TLS client for the etcd cluster can privilege escalate by becoming a full etcd cluster member. If we used different certificate authorities for clients and peers, then it would be possible to enable etcd RBAC, and create a TLS cert with a specific user as the CN that only has access to the /metrics port.

**Comment by randomvariable** [comment]
Another way you could set this up, but would be a bit more secure would be to run https://github.com/brancz/kube-rbac-proxy as a daemonset on the control plane nodes. Configure this to load up the existing etcd certificates on disk, and then only allow /metrics to be forwarded to the etcd port. 

It should then be possible to set up a Service pointing at the DaemonSet with a ServiceMonitor using Prometheus' own Kubernetes service account as the bearer token.

**Comment by pnordh** [comment]
We resolved to use the `postKubeadmCommands ` directive to create a Secret in the target cluster with the etcd healthcheck certificate generated by kubeadm. 

As said above though, that certificate is valid for peer connections unfortunately.

**Comment by randomvariable** [comment]
I know we don't have a label for it, but just for tracking

/area node-agent

**Comment by tuxtof** [comment]
> We resolved to use the `postKubeadmCommands` directive to create a Secret in the target cluster with the etcd healthcheck certificate generated by kubeadm.
> 
> As said above though, that certificate is valid for peer connections unfortunately.

@pnordh any example to share ?
